File: 13_回溯算法/经典bt/hukumenzan-覆面算.py
# https://maspypy.github.io/library/other/solve_hukumenzan.hpp
# 覆面算(口算难题/竖式计算)
# a + b = c を解く

from typing import List, Tuple


def solveHukumenzan(A: str, B: str, C: str) -> List[Tuple[int, int, int]]:
    res = []
    v = list(set(list(A + B + C)))
    if len(v) > 10:
        return res
    order = list(range(10))
    while True:
        ok, order = nextPermutation(order, inPlace=True)
        if not ok:
            break
        mp = {v[i]: order[i] for i in range(len(v))}
        if mp[A[0]] == 0 or mp[B[0]] == 0 or mp[C[0]] == 0:
            continue
        a, b, c = 0, 0, 0
        for x in A:
            a = 10 * a + mp[x]
        for x in B:
            b = 10 * b + mp[x]
        for x in C:
            c = 10 * c + mp[x]
        if a + b == c:
            res.append((a, b, c))
    return res


from typing import Any, MutableSequence, Tuple


def nextPermutation(nums: MutableSequence[Any], inPlace=False) -> Tuple[bool, MutableSequence[Any]]:
    """返回下一个字典序的排列，如果不存在，返回本身;时间复杂度O(n)"""
    if not inPlace:
        nums = nums[:]

    left = right = len(nums) - 1

    while left > 0 and nums[left - 1] >= nums[left]:  # 1. 找到最后一个递增位置
        left -= 1
    if left == 0:  # 全部递减
        return False, nums
    last = left - 1  # 最后一个递增位置

    while nums[right] <= nums[last]:  # 2. 找到最小的可交换的right，交换这两个数
        right -= 1
    nums[last], nums[right] = nums[right], nums[last]

    left, right = last + 1, len(nums) - 1  # 3. 翻转后面间这段递减数列
    while left < right:
        nums[left], nums[right] = nums[right], nums[left]
        left += 1
        right -= 1
    return True, nums


# https://atcoder.jp/contests/abc198/tasks/abc198_d
if __name__ == "__main__":
    # A,B,C都是小写字母组成的长度不超过10的字符串

    A, B, C = input(), input(), input()
    res = solveHukumenzan(A, B, C)
    if not res:
        print("UNSOLVABLE")
    else:
        a, b, c = res[0]
        print(a)
        print(b)
        print(c)

File: 13_回溯算法/经典bt/hukumenzan-覆面算.go
package main

import (
	"bufio"
	"fmt"
	"os"
)

func main() {
	in := bufio.NewReader(os.Stdin)
	out := bufio.NewWriter(os.Stdout)
	defer out.Flush()

	var A, B, C string
	fmt.Fscan(in, &A, &B, &C)
	res := solveHukumenzan(A, B, C)
	if len(res) == 0 {
		fmt.Fprintln(out, "UNSOLVABLE")
		return
	}
	a, b, c := res[0][0], res[0][1], res[0][2]
	fmt.Fprintln(out, a, b, c)
}

func solveHukumenzan(A, B, C string) [][]int {
	res := [][]int{}
	set := make(map[byte]struct{})
	for _, v := range A + B + C {
		set[byte(v)] = struct{}{}
	}
	if len(set) > 10 {
		return res
	}
	chars := make([]byte, 0, len(set))
	for k := range set {
		chars = append(chars, k)
	}

	order := make([]int, 10)
	for i := range order {
		order[i] = i
	}

	for order, ok := nextPermutation(order, true); ok; order, ok = nextPermutation(order, true) {
		mp := make(map[byte]int)
		for i := range chars {
			mp[chars[i]] = order[i]
		}
		if mp[A[0]] == 0 || mp[B[0]] == 0 || mp[C[0]] == 0 {
			continue
		}
		a, b, c := 0, 0, 0
		for i := 0; i < len(A); i++ {
			a = 10*a + mp[A[i]]
		}
		for i := 0; i < len(B); i++ {
			b = 10*b + mp[B[i]]
		}
		for i := 0; i < len(C); i++ {
			c = 10*c + mp[C[i]]
		}

		if a+b == c {
			res = append(res, []int{a, b, c})
		}
	}

	return res
}

// 返回下一个字典序的排列.
func nextPermutation(nums []int, inPlace bool) (res []int, ok bool) {
	if !inPlace {
		nums = append(nums[:0:0], nums...)
	}
	left, right := len(nums)-1, len(nums)-1
	for left > 0 && nums[left-1] >= nums[left] {
		left--
	}
	if left == 0 {
		return
	}
	last := left - 1
	for nums[right] <= nums[last] {
		right--
	}
	nums[last], nums[right] = nums[right], nums[last]
	for i, j := last+1, len(nums)-1; i < j; i, j = i+1, j-1 {
		nums[i], nums[j] = nums[j], nums[i]
	}
	return nums, true
}
